[Yi Kai] iP

kai824 opened the pull request:
# Chatbot Adam
> "Use GFMD in the PR description" 😔 -[Course Site](https://nus-cs2103-ay2425s2.github.io/website/schedule/week4/project.html#1-use-gfmd-in-the-pr-description)

Uses:

- managing various types of tasks
- super ~hard~ **EASY** to learn and use 🤩 

Types of task that can be handled:
1. Deadline tasks
2. Events with a dead range
3. Generic to-dos

Feature to-do list:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Editing deadlines
- [ ] Reminders

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

kai824 committed:
Update gitignore to ignore .class files

kai824 committed:
Implement chatbot skeleton

kai824 committed:
Rename chatbot to Adam

kai824 committed:
Implement echo feature

kai824 committed:
Implement Add, List feature

kai824 committed:
Implement tasks with marking as done or undone

kai824 committed:
Store indentation separately

kai824 committed:
Implement types of tasks

kai824 committed:
Implement classes for types of tasks

kai824 committed:
Update text files for UI testing

kai824 committed:
Update windows file for testing

kai824 committed:
Implement exceptions to be thrown and handled

kai824 committed:
Add throwing and handling of various errors

It was also decided that some arguments should be allowed to be empty
for now, so EmptyDescription only handles empty descriptions rather than
  empty arguments.

kai824 committed:
Fix critical bug of Event to being null

kai824 committed:
Update tests and handle mark negative indexes

kai824 committed:
Add deletion functionality

kai824 committed:
Add tests for handling task deletion

kai824 committed:
Fix formatting issues

kai824 committed:
Update variable name to follow coding standard

kai824 committed:
Refactor task management to a different class

kai824 committed:
Save the tasklist to logfile on every change

kai824 committed:
Update gitignore to ignore tasklist files

kai824 committed:
Load tasklist from file

kai824 committed:
Add tests for loading from log file

kai824 committed:
Merge branch-Level-7 into master

kai824 committed:
Store dates as LocalDate and not Strings

kai824 committed:
Update tests for date formatting

kai824 committed:
Handle invalid date input

kai824 committed:
Add command to list events on date

kai824 committed:
Update tests for listing on specific date

kai824 committed:
Merge branch 'branch-Level-8': date features

kai824 committed:
Separate classes for Ui, TaskList, Storage

kai824 committed:
Add Parser class

kai824 committed:
Refactor Adam.java and create Command classes

kai824 committed:
Merge branch 'more-oop'

kai824 committed:
Reorganise files

kai824 committed:
Fix all package issues

kai824 committed:
Merge branch 'packages'

kai824 committed:
Merge branch 'add-gradle-support'

kai824 committed:
Fix typo

kai824 committed:
Merge branch 'packages'

kai824 committed:
Modify gradle setting

kai824 committed:
Add JUnit tests for Parser and TaskList

kai824 committed:
Merge branch 'branch-A-JUnit'

kai824 committed:
Add find command

kai824 committed:
Update text tests

kai824 committed:
Update error message

kai824 committed:
Fix formatting and unused variables

kai824 committed:
Fix boolean variable names

kai824 committed:
Add javadoc for most classes and methods

kai824 committed:
Add javadoc for test files

kai824 committed:
Merge branch 'branch-Level-9'

kai824 committed:
Merge branch 'branch-A-CodingStandard'

kai824 committed:
Merge branch 'branch-A-javadoc'

kai824 committed:
Add checkstyle feature with gradle

kai824 committed:
Fix checkstyle issues found

kai824 committed:
Fix package name warning

kai824 committed:
Merge branch 'branch-A-CheckStyle'

Shanyey commented on src/main/java/adam/core/Adam.java:
```diff
@@ -0,0 +1,55 @@
+    /** An instance of Ui handling user interaction */
+    private Ui ui;
+
+    Adam() {
```
Is there a reason for your constructors to not have a public access modifier?

Shanyey commented on src/main/java/adam/parser/Command.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * Represents a command that can be executed by the user.
+ */
+public abstract class Command {
```
Can consider adding a new package calleed adam.command instead of combining the parser and command classes.

Shanyey commented on src/main/java/adam/parser/DeleteCommand.java:
```diff
@@ -0,0 +1,48 @@
+    public void execute(TaskList manager, Ui ui) throws AdamException {
+        try {
+            Task task = manager.deleteTask(this.index);
+            ui.outputText("OK, I've deleted this task:");
```
Might be more concise to extract the messages into the ui class and simply passing in the values to a method like ui.printDeletedText(task)

Shanyey commented on src/main/java/adam/tasks/ToDo.java:
```diff
@@ -0,0 +1,41 @@
+     * Gets the task as a String for logging.
+     */
+    @Override
+    public String log() {
```
can rename methods to a noun like getLog()

Shanyey reviewed (commented):
Overall very nice code structure and standards. Good use of javadoc comments.
Can consider shortening parse by extracting out the print statements.

senshir commented on src/main/java/adam/parser/Parser.java:
```diff
@@ -0,0 +1,81 @@
+     * @param date The date to convert.
+     * @return The date as a string in the format dd MMM yyyy.
+     */
+    public static String toOutputDate(LocalDate date) {
```
toOutputDate and toLogDate sound okay as method names, but maybe you could consider using verbs like formatOutputDate/formatLogDate instead? It could help to indicate the action done clearly.
Overall great use of method and variable naming

senshir commented on src/main/java/adam/parser/Parser.java:
```diff
@@ -0,0 +1,81 @@
+            return new UnmarkCommand(input);
+        } else if (DeleteCommand.isMatch(input)) {
+            return new DeleteCommand(input);
+        } else if (FindCommand.matches(input)) {
```
May be more concise if FindCommand also had a method called isMatch instead of matches?

senshir commented on src/main/java/adam/tasks/Task.java:
```diff
@@ -0,0 +1,172 @@
+        } catch (IndexOutOfBoundsException e) {
+            throw new InvalidLogFile();
+        }
+        if (parts.get(1) == "true") {
```
May want to consider using .equals for a string instead of == as it means object reference.

senshir commented on src/main/java/adam/tasks/Task.java:
```diff
@@ -0,0 +1,172 @@
+                delimiterIndex++;
+            } else {
+                if (cur.length() > 0) {
+                    cur += " ";
```
A bit inefficient to use string concatenation by adding them. Could be better to use StringBuilder and append?

senshir commented on src/main/java/adam/tasks/Task.java:
```diff
@@ -0,0 +1,172 @@
+     *
+     * @return the description to be logged.
+     */
+    public String log() {
```
I think in this case toLogString() would be better for naming conventions, especially since log is a noun.

senshir reviewed (commented):
Overall really good use of OOP and methods. Just some naming conventions and efficiency stuff that you might wanna look at.

kai824 committed:
Change CLI to GUI

kai824 committed:
Add greeting at the start

kai824 committed:
Add greeting and close window on Bye command

kai824 committed:
Merge branch 'branch-Level-10'

kai824 committed:
Add assertions into the code where appropriate

The current codebase does not use assertions to prevent unexpected
behaviour. This commit adds assertions in a few edge cases which are not
already caught by exceptions, such as:
* checking that the "done" field must either be "true" or "false" when
  loading from a log file
* checking that any Exception messages are not empty

kai824 committed:
Refactor Command classes into a new package

Previously, all Command classes were in a same package as the Parser
class. While these classes all deal with parsing input, with the growing
number of Command classes, it would be more sensible to put all the
Command classes into a different package.

Hence, this commit refactors them into a different class, while making
the necessary changes to the import statements.

kai824 committed:
Rename function names based on coding standard

Renamed various functions based on coding standard:
* Parser::toLogDate -> formatLogDate
* Parser::toOutputDate -> formatOutputdate
* Task::log -> toLogString

These function names should be more informative about their expected
behaviour.

kai824 committed:
Add tests for package Tasks

Among all remaining classes that have no tests yet, Tasks is the most
crucial to be tested on. As such, this commit adds tests for the
following classes:
* Deadline
* Event
* ToDo
* and Task itself, to test some methods which all child classes inherit.

kai824 committed:
Merge pull request #1 from kai824/branch-A-Assertions

Add assertions into the code where appropriate

kai824 committed:
Merge pull request #2 from kai824/branch-A-CodeQuality

Branch code quality

kai824 committed:
Add fixed duration tasks

kai824 committed:
Add new exception for invalid duration

kai824 committed:
Merge pull request #3 from kai824/B-FixedDurationTasks

Fixed duration tasks

kai824 committed:
Update user guide

kai824 committed:
Add assertion that Event starts before it ends

kai824 committed:
Merge pull request #5 from kai824/branch-A-UserGuide

Add user guide
